File: src/clj/big_config/git.clj
(ns big-config.git
  (:require
   [big-config :as bc]
   [big-config.core :refer [->workflow]]
   [big-config.run :refer [generic-cmd]]))

(defn get-revision [revision key opts]
  (let [revision (cond
                   (string? revision) revision
                   (keyword? revision) (revision opts)
                   :else (throw (ex-info "Revision is neither a string nor a keyword" {:revision revision
                                                                                       :key key
                                                                                       :opts opts})))
        cmd (format "git rev-parse %s" revision)]
    (generic-cmd :opts opts
                 :cmd cmd
                 :key key)))

(defn fetch-origin [opts]
  (generic-cmd :opts opts
               :cmd "git fetch origin"))

(defn upstream-name [key opts]
  (let [cmd "git rev-parse --abbrev-ref @{upstream}"]
    (generic-cmd :opts opts
                 :cmd cmd
                 :key key)))

(defn git-diff [opts]
  (generic-cmd :opts opts
               :cmd "git diff --quiet"))

(defn git-push [opts]
  (generic-cmd :opts opts
               :cmd "git push"))

(defn compare-revisions [opts]
  (let [{:keys [::prev-revision
                ::current-revision
                ::origin-revision]} opts
        res (or (= prev-revision origin-revision)
                (= current-revision origin-revision))]
    (merge opts (if res
                  {::bc/exit 0
                   ::bc/err nil}
                  {::bc/exit 1
                   ::bc/err "The local revisions don't match the remote revision"}))))

(def check (->workflow {:first-step ::git-diff
                        :wire-fn (fn [step _]
                                   (case step
                                     ::git-diff [git-diff ::fetch-origin]
                                     ::fetch-origin [fetch-origin ::upstream-name]
                                     ::upstream-name [(partial upstream-name ::upstream-name) ::pre-revision]
                                     ::pre-revision [(partial get-revision "HEAD~1" ::prev-revision) ::current-revision]
                                     ::current-revision [(partial get-revision "HEAD" ::current-revision) ::origin-revision]
                                     ::origin-revision [(partial get-revision ::upstream-name ::origin-revision) ::compare-revisions]
                                     ::compare-revisions [compare-revisions ::end]
                                     ::end [identity]))}))

(comment
  (->> (check {})
       (into (sorted-map))))

File: src/git.ts
import { createWorkflow } from "./core.js";
import { ERR, EXIT, GIT_CURRENT_REVISION, GIT_ORIGIN_REVISION, GIT_PREV_REVISION, GIT_UPSTREAM_NAME, type Opts } from "./keys.js";
import { genericCmd } from "./run.js";
import { registerFunction } from "./utils.js";

export function getRevision(revision: string, key: string, opts: Opts): Opts {
  const resolved = Object.prototype.hasOwnProperty.call(opts, revision) ? opts[revision] : revision;
  if (typeof resolved !== "string") throw Object.assign(new Error("Revision is neither a string nor a keyword"), { data: { revision, key, opts } });
  return genericCmd({ opts, cmd: ["git", "rev-parse", resolved], key });
}

export function fetchOrigin(opts: Opts): Opts {
  return genericCmd({ opts, cmd: ["git", "fetch", "origin"] });
}

export function upstreamName(key: string, opts: Opts): Opts {
  return genericCmd({ opts, cmd: ["git", "rev-parse", "--abbrev-ref", "@{upstream}"], key });
}

export function gitDiff(opts: Opts): Opts {
  return genericCmd({ opts, cmd: ["git", "diff", "--quiet"] });
}

export function gitPush(opts: Opts): Opts {
  return genericCmd({ opts, cmd: ["git", "push"] });
}

export function compareRevisions(opts: Opts): Opts {
  const res = opts[GIT_PREV_REVISION] === opts[GIT_ORIGIN_REVISION] || opts[GIT_CURRENT_REVISION] === opts[GIT_ORIGIN_REVISION];
  return res
    ? { ...opts, [EXIT]: 0, [ERR]: null }
    : { ...opts, [EXIT]: 1, [ERR]: "The local revisions don't match the remote revision" };
}

export const check = createWorkflow({
  firstStep: "big-config.git/git-diff",
  wireFn: (step) => {
    switch (step) {
      case "big-config.git/git-diff": return [gitDiff, "big-config.git/fetch-origin"];
      case "big-config.git/fetch-origin": return [fetchOrigin, "big-config.git/upstream-name"];
      case "big-config.git/upstream-name": return [(opts) => upstreamName(GIT_UPSTREAM_NAME, opts), "big-config.git/pre-revision"];
      case "big-config.git/pre-revision": return [(opts) => getRevision("HEAD~1", GIT_PREV_REVISION, opts), "big-config.git/current-revision"];
      case "big-config.git/current-revision": return [(opts) => getRevision("HEAD", GIT_CURRENT_REVISION, opts), "big-config.git/origin-revision"];
      case "big-config.git/origin-revision": return [(opts) => getRevision(GIT_UPSTREAM_NAME, GIT_ORIGIN_REVISION, opts), "big-config.git/compare-revisions"];
      case "big-config.git/compare-revisions": return [compareRevisions, "big-config.git/end"];
      case "big-config.git/end": return [(x) => x];
      default: return [(x) => x];
    }
  }
});

registerFunction("big-config.git/check", check as any);
registerFunction("big-config.git/git-push", gitPush);

export { getRevision as "get-revision", fetchOrigin as "fetch-origin", upstreamName as "upstream-name", gitDiff as "git-diff", gitPush as "git-push", compareRevisions as "compare-revisions" };
